packages/ui-components/src/Button/Button.tsx

```diff
@@ -1,50 +1,70 @@
-import React, { FunctionComponent, ButtonHTMLAttributes } from "react";
+import React, { ElementType } from "react";
 import classnames from "classnames/bind";
 
 import objectToClassnames from "../utils/objectToClassnames";
 
 import styles from "./Button.module.scss";
 
 export type ButtonIntent =
-  | "primary-success"
+  | "primary"
   | "primary-danger"
   | "secondary"
   | "tertiary";
+
 export type ButtonSize = "standard" | "small";
 
-type OwnButtonProps = {
+type ElementProps<Type extends ElementType> = Pick<
+  React.ComponentProps<Type>,
+  keyof React.ComponentProps<Type>
+>;
+
+type CombinedProps<OwnProps, NativeElementProps> = OwnProps &
+  Omit<NativeElementProps, keyof OwnProps>;
+
+type OwnButtonProps<Element> = {
+  as: Element;
   intent?: ButtonIntent;
   size?: ButtonSize;
-  onClick?: () => void;
+  fluid?: boolean;
 };
 
-type NativeButtonProps = Omit<
-  ButtonHTMLAttributes<HTMLButtonElement>,
-  keyof OwnButtonProps
+type ButtonProps<OwnProps, Type extends ElementType> = CombinedProps<
+  OwnProps,
+  ElementProps<Type>
```

Comment:
I don't think you need `ElementProps` since its definition might be redundant. You can use `React.ComponentProps<Type>` directly since your definition of `ElementProps` picks all the keys from `React.ComponentProps`. If you want to omit some standard ones there appear to be helpers like `React.ComponentPropsWithoutRef<Type>`. Not sure what applies in this scenario.